Add ability to inherit debugger providers from global config

chamilad opened the pull request:
This changes the implementation to inherit and override global debugger adapter configuration instead of having to redefine it in the project specific config file. Since some adapters need the absolute path to the binary having it in the project local file might not be portable.

> initial implementation of replacing adapter configuration has been changed to overriding.

This change will not break existing config.

## Example
### `init.lua`
```lua filename="init.lua"
local dap = require('dap')
dap.adapters.cppdbg = {
-- global debug adapter
}

dap.configurations.rust = {
-- global debug config
}

local dap_projects = require("nvim-dap-projects")
dap_projects.search_project_config()
```

### `<project-root>/project-config.lua`
```lua filename="project-config.lua"
local dap = require('dap')
-- optional adapter configuration to be defined here
-- adapter configuration defined here will be merged
-- with the global config in init.lua

dap.configurations.rust = {
-- project specific config
}
```

## Why?
I found that certain debug config (in my case it was Rust debugging with `cppdbg`) required defining absolute path names in the adapter config. Having this detail in the project specific config is brittle since if the repo is shared, each developer will have to change that value. Git ignoring the file defeats the purpose too IMO.

## Drawbacks
> initial implementation had drawback of replacing config, but this has been updated.

chamilad committed:
Add ability to inherit debugger providers from global config

This change adds a setup config option to inherit global debugger
adapter configuration instead of having to redefine it in the project
specific config file. Since some adapters need the absolute path to the
binary having it in the project local file might not be portable.

This change adds the inherit_adapters setup option, by default set to
false. Setting it to false will keep the current behavior, which is to
totally ignore the global nvim-dap configuration. Setting it to true,
which should be explicitly done,  will not read the project local
adapters section, instead using the global section.

This change will not break existing config since setup() function is
optional and defaults work to override adapters.

todo: merge global and local adapter sections instead of replacing.

ldelossa commented:
Hey. 

Cool idea. 

My immediate concern with this is, things may get a little confusing. 

When a project has a local config file which also provides adapters, this should take precedence over the global adapters defined in nvim-dap configuration. 

Id almost rather this become the default case. 

Do you want to modify this PR such that, we add adapters to nvim-dap (not just clear the inventory), and if an adapter is defined in both global config and project config, the project one takes precedence. Would this also work for your case?

chamilad commented:
That's a great idea. In my tunnel vision I thought either-or. 

Let me modify this and test and push another commit. Thanks!!

chamilad committed:
Change design to add and override adapters

Incorporating suggestion from the PR to add to adapter configuration and
override if the same config exists.

chamilad commented:
I manually tested this change with
1. global adapter config, no local adapter config
2. global adapter config, local adapter config with a different adapter name (merge adapters scenario)
3. global adapter config, local adapter config with the same name (override adapters scenario)

Have I missed anything?
